coordinator/statement_executor.go

```diff
@@ -606,6 +517,137 @@ func (e *StatementExecutor) executeSelectStatement(stmt *influxql.SelectStatemen
 	return nil
 }
 
+func (e *StatementExecutor) createIterators(stmt *influxql.SelectStatement, ctx *influxql.ExecutionContext) ([]influxql.Iterator, *influxql.SelectStatement, error) {
+	// Handle SHOW MEASUREMENTS at the database level instead of delegating it to the shards.
+	if source, ok := stmt.Sources[0].(*influxql.Measurement); ok && source.Name == "_measurements" {
+		// Use the optimized version only if we have direct access to the database.
+		if store, ok := e.TSDBStore.(LocalTSDBStore); ok {
+			index := store.DatabaseIndex(source.Database)
+			if index == nil {
+				return nil, stmt, nil
+			}
+
+			opt := influxql.IteratorOptions{
+				Condition: stmt.Condition,
+				Aux:       []influxql.VarRef{{Val: "name", Type: influxql.String}},
+				Limit:     stmt.Limit,
+				Offset:    stmt.Offset,
+			}
+
+			var input influxql.Iterator
+			input, err := tsdb.NewMeasurementIterator(index, opt)
+			if err != nil {
+				return nil, stmt, err
+			}
+
+			// Apply limit & offset.
+			if opt.Limit > 0 || opt.Offset > 0 {
+				input = influxql.NewLimitIterator(input, opt)
+			}
+
+			aitr := influxql.NewAuxIterator(input, opt)
+			itr := aitr.Iterator("name", influxql.String)
+			aitr.Background()
+			return []influxql.Iterator{itr}, stmt, nil
+		}
```

Comment:
Do you mean extra the double nested `if` to its own function or extract the contents inside of the double nested `if` to another function?